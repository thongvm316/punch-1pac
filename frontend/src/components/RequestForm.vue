<template>
  <div>
    <div
      class="form-group"
      :class="{ 'has-error': $v.day.$error }"
    >
      <label class="form-label">{{ $t('label.date') }}</label>
      <flat-pickr
        v-model="$v.day.$model"
        :config="{ enable: [day], locale: flatpickrLocaleMapper[pickrLocale] }"
        class="form-input daterange-picker"
      />
      <p
        v-if="$v.day.$error"
        class="form-input-hint"
      >
        {{ $t('validation.required', { name: $t('label.date') }) }}
      </p>
    </div>
    <div
      class="form-group"
      :class="{ 'has-error': $v.params.attended_at.$error || errors.attended_at }"
    >
      <label class="form-label">{{ $t('label.attendedAt') }}</label>
      <flatPickr
        v-model="$v.params.attended_at.$model"
        class="form-input time-picker"
        :config="{
          enableTime: true,
          noCalendar: true,
          dateFormat: 'H:i',
          time_24hr: true}"
      />
      <p
        v-if="$v.params.attended_at.$error && !errors.attended_at"
        class="form-input-hint"
      >
        {{ $t('validation.required', { name: $t('label.attendedAt') }) }}
      </p>
      <p
        v-if="errors.attended_at"
        class="form-input-hint"
      >
        {{ $t('label.attendedAt') }} {{ errors.attended_at[0] }}
      </p>
    </div>
    <div
      class="form-group"
      :class="{ 'has-error': $v.params.left_at.$error || errors.left_at }"
    >
      <label class="form-label">{{ $t('label.leftAt') }}</label>
      <flatPickr
        v-model="$v.params.left_at.$model"
        class="form-input time-picker"
        :config="{
          enableTime: true,
          noCalendar: true,
          dateFormat: 'H:i',
          time_24hr: true}"
      />
      <p
        v-if="$v.params.left_at.$error && !errors.left_at"
        class="form-input-hint"
      >
        {{ $t('validation.required', { name: $t('label.leftAt') }) }}
      </p>
      <p
        v-if="errors.left_at"
        class="form-input-hint"
      >
        {{ $t('label.leftAt') }} {{ errors.left_at[0] }}
      </p>
    </div>
    <div
      class="form-group"
      :class="{ 'has-error': $v.params.reason.$error || errors.reason }"
    >
      <label class="form-label">{{ $t('label.reason') }}</label>
      <textarea
        v-model="$v.params.reason.$model"
        class="form-input"
      />
      <p
        v-if="$v.params.reason.$error && !errors.reason"
        class="form-input-hint"
      >
        {{ $t('validation.required', { name: $t('label.reason') }) }}
      </p>
      <p
        v-if="errors.reason"
        class="form-input-hint"
      >
        {{ $t('label.reason') }} {{ errors.reason[0] }}
      </p>
    </div>
    <div class="form-group">
      <button
        v-if="attendance"
        ref="localAddRequestButton"
        type="button"
        class="btn btn-success btn-submit"
        :disabled="isDisabled"
        @click="localAddRequest"
      >
        {{ $t('button.common.add') }}
      </button>
      <button
        v-else
        ref="localEditRequestButton"
        type="button"
        class="btn btn-success btn-submit"
        :disabled="isDisabled"
        @click="localEditRequest"
      >
        {{ $t('button.common.save') }}
      </button>
    </div>
  </div>
</template>

<script>
import flatpickrLocale from '../mixins/flatpickr-locale'
import handleSuccess from '../mixins/handle-success'
import requestFormValidate from '../validations/request-form-valiate'
import { CLEAR_REQUEST_ERRORS } from '../store/mutation-types'
import { mapState, mapActions, mapMutations } from 'vuex'
import { isEmpty } from 'underscore'
const flatPickr = () => import('vue-flatpickr-component')

export default {
  name: 'RequestForm',

  components: {
    flatPickr
  },

  mixins: [flatpickrLocale, handleSuccess, requestFormValidate],

  props: {
    attendance: Object,
    request: Object
  },

  data() {
    return {
      day: '',
      params: {
        attendance_day: '',
        attended_at: '',
        left_at: '',
        reason: ''
      }
    }
  },

  computed: {
    ...mapState('requests', ['errors'])
  },

  methods: {
    ...mapActions('requests', ['addRequest', 'updateRequest']),

    ...mapMutations('requests', [CLEAR_REQUEST_ERRORS]),

    localAddRequest() {
      this.addRequest(this.params).then(response => {
        this.handleSuccess({
          emitType: 'afterModify',
          message: this.$t('messages.request.createSuccess')
        })
      })
    },

    localEditRequest() {
      this.updateRequest({ id: this.request.id, params: this.params }).then(response => {
        this.handleSuccess({
          emitType: 'afterModify',
          message: this.$t('messages.request.updateSuccess')
        })
      })
    }
  },

  created() {
    if (!isEmpty(this.errors)) this[CLEAR_REQUEST_ERRORS]()

    if (this.attendance) {
      this.day = this.params.attendance_day = this.attendance.day
      const statuses = ['attended_at', 'left_at']
      statuses.forEach(key => {
        this.params[key] = this.attendance[key]
      })
    } else if (this.request) {
      this.day = this.request.attendance_day
      this.params = { ...this.request }
    }
  }
}
</script>
