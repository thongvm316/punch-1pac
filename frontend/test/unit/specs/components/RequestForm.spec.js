import { shallowMount } from '@vue/test-utils'
import wrapperOps from '../../supports/wrapper'
import attendancesData from '../../supports/fixtures/attendances.api'
import requestsData from '../../supports/fixtures/requests.api'
import RequestForm from '@/components/RequestForm'

const localAddRequest = jest.spyOn(RequestForm.methods, 'localAddRequest')
const localEditRequest = jest.spyOn(RequestForm.methods, 'localEditRequest')
const addRequest = jest.spyOn(RequestForm.methods, 'addRequest').mockResolvedValue(null)
const updateRequest = jest.spyOn(RequestForm.methods, 'updateRequest').mockResolvedValue(null)
const handleSuccess = jest.spyOn(RequestForm.mixins[1].methods, 'handleSuccess')

const attendance = [...attendancesData.attendances][0]
const request = [...requestsData.requests][0]
const errors = {
  attended_at: ['attended_at wrong'],
  left_at: ['left_at wrong']
}

describe('RequestForm.vue', () => {
  let wrapper

  afterEach(() => { wrapper.vm.$destroy() })

  describe('when have no props', () => {
    const localWrapperOps = {
      ...wrapperOps,
      stubs: {
        flatPickr: true
      }
    }

    beforeEach(() => {
      wrapper = shallowMount(RequestForm, localWrapperOps)
    })

    it('should render correctly', () => {
      expect(wrapper.exists()).toBeTruthy()
      expect(wrapper.isVueInstance()).toBeTruthy()
      expect(wrapper).toMatchSnapshot()
    })
  })


  describe('when attendance props', () => {
    const localWrapperOps = {
      ...wrapperOps,
      propsData: { attendance },
      stubs: {
        flatPickr: true
      }
    }

    beforeEach(() => {
      wrapper = shallowMount(RequestForm, localWrapperOps)
    })

    it('should render correctly', () => {
      expect(wrapper.exists()).toBeTruthy()
      expect(wrapper.isVueInstance()).toBeTruthy()
      expect(wrapper).toMatchSnapshot()
    })

    it('should localAddRequest toHaveBeenCalled', async () => {
      wrapper.find({ ref: 'localAddRequestButton' }).trigger('click')
      await wrapper.vm.$nextTick()

      expect(localAddRequest).toHaveBeenCalled()
      expect(addRequest).toHaveBeenCalledWith(wrapper.vm.params)
      expect(handleSuccess).toHaveBeenCalledWith({ emitType: 'afterModify', message: 'Request is created' })
    })
  })

  describe('when request props', () => {
    const localWrapperOps = {
      ...wrapperOps,
      propsData: { request },
      stubs: {
        flatPickr: true
      },
      computed: {
        isDisabled() {
          return false
        }
      }
    }

    beforeEach(() => {
      wrapper = shallowMount(RequestForm, localWrapperOps)
    })

    it('should render correctly', () => {
      expect(wrapper.exists()).toBeTruthy()
      expect(wrapper.isVueInstance()).toBeTruthy()
      expect(wrapper).toMatchSnapshot()
    })

    it('should localEditRequest toHaveBeenCalled', async () => {
      wrapper.find({ ref: 'localEditRequestButton' }).trigger('click')
      await wrapper.vm.$nextTick()

      expect(localEditRequest).toHaveBeenCalled()
      expect(updateRequest).toHaveBeenCalledWith({ id: request.id, params: wrapper.vm.params })
      expect(handleSuccess).toHaveBeenCalledWith({ emitType: 'afterModify', message: 'Request is updated' })
    })
  })

  describe('when errors', () => {
    const localWrapperOps = {
      ...wrapperOps,
      propsData: { request },
      stubs: {
        flatPickr: true
      },
      computed: {
        errors() {
          return errors
        }
      }
    }

    beforeEach(() => {
      wrapper = shallowMount(RequestForm, localWrapperOps)
    })

    it('should render errors', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })
})
