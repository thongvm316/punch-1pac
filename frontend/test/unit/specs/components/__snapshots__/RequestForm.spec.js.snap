// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestForm.vue when attendance props should render correctly 1`] = `
<div>
  <div class="form-group"><label class="form-label">Date</label>
    <flatpickr-stub config="[object Object]" value="2020-01-01" class="form-input daterange-picker"></flatpickr-stub>
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Attended At</label>
    <flatpickr-stub config="[object Object]" value="09:45" class="form-input time-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Left At</label>
    <flatpickr-stub config="[object Object]" value="09:45" class="form-input time-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Reason</label> <textarea class="form-input"></textarea>
    <!---->
    <!---->
  </div>
  <div class="form-group"><button type="button" class="btn btn-success btn-submit">
      Add
    </button></div>
</div>
`;

exports[`RequestForm.vue when errors should render errors 1`] = `
<div>
  <div class="form-group"><label class="form-label">Date</label>
    <flatpickr-stub config="[object Object]" value="2020-02-07" class="form-input daterange-picker"></flatpickr-stub>
    <!---->
  </div>
  <div class="form-group has-error"><label class="form-label">Attended At</label>
    <flatpickr-stub config="[object Object]" value="10:48" class="form-input time-picker"></flatpickr-stub>
    <!---->
    <p class="form-input-hint">
      Attended At attended_at wrong
    </p>
  </div>
  <div class="form-group has-error"><label class="form-label">Left At</label>
    <flatpickr-stub config="[object Object]" value="10:48" class="form-input time-picker"></flatpickr-stub>
    <!---->
    <p class="form-input-hint">
      Left At left_at wrong
    </p>
  </div>
  <div class="form-group"><label class="form-label">Reason</label> <textarea class="form-input"></textarea>
    <!---->
    <!---->
  </div>
  <div class="form-group"><button type="button" disabled="disabled" class="btn btn-success btn-submit">
      Save
    </button></div>
</div>
`;

exports[`RequestForm.vue when have no props should render correctly 1`] = `
<div>
  <div class="form-group"><label class="form-label">Date</label>
    <flatpickr-stub config="[object Object]" value="" class="form-input daterange-picker"></flatpickr-stub>
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Attended At</label>
    <flatpickr-stub config="[object Object]" value="" class="form-input time-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Left At</label>
    <flatpickr-stub config="[object Object]" value="" class="form-input time-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Reason</label> <textarea class="form-input"></textarea>
    <!---->
    <!---->
  </div>
  <div class="form-group"><button type="button" class="btn btn-success btn-submit">
      Save
    </button></div>
</div>
`;

exports[`RequestForm.vue when request props should render correctly 1`] = `
<div>
  <div class="form-group"><label class="form-label">Date</label>
    <flatpickr-stub config="[object Object]" value="2020-02-07" class="form-input daterange-picker"></flatpickr-stub>
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Attended At</label>
    <flatpickr-stub config="[object Object]" value="10:48" class="form-input time-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Left At</label>
    <flatpickr-stub config="[object Object]" value="10:48" class="form-input time-picker"></flatpickr-stub>
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Reason</label> <textarea class="form-input"></textarea>
    <!---->
    <!---->
  </div>
  <div class="form-group"><button type="button" class="btn btn-success btn-submit">
      Save
    </button></div>
</div>
`;
